### PR TITLE
[SFA-82] Update users collection on login/registration

### PR DESCRIPTION
**Pull request by cyamonide**
## Description

This PR adds new (and existing) users to the users collection whenever they either log in or register for a new account. Also necessitated was a method to create/update users in the Firestore collection, hence the addition of a `.upsert` method.

This PR also changes the profile page to display users' real names instead of their UID.

### CODE REVIEW

File: app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt [hunk 1]
```diff
@@ -32,6 +32,18 @@ class FirebaseUserRepository(
             .mapNotNull { document -> document.toUserEntity() }
             .toDomain()
     }
+
+    override fun upsert(_user: User): User {
+        val user = _user.toEntity()
+        usersCollection.document(user.id).set(user)
```
Comment (aaronchlam) [inline app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt]: can't use await here?

File: app/src/main/java/com/uwcs446/socialsports/services/user/current/FirebaseCurrentUserRepository.kt [hunk 2]
```diff
@@ -32,6 +35,7 @@ class FirebaseCurrentUserRepository
 
     override fun handleAuthChange() {
         updateCurrentUser()
+        upsertUserCollection()
```
Comment (aaronchlam) [inline app/src/main/java/com/uwcs446/socialsports/services/user/current/FirebaseCurrentUserRepository.kt:38]: Can we do this only if the user is registered? Are we able to get the auth change type?

Comment (cyamonide) [reply app/src/main/java/com/uwcs446/socialsports/services/user/current/FirebaseCurrentUserRepository.kt:38]: It doesn't look like we can get the auth change type, at least from the docs.

https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseAuth.AuthStateListener

`onAuthStateChanged` in `MainActivity.kt` calls `handleAuthChange`, and the only argument passed to `onAuthStateChanged` is a `FirebaseAuth` instance, which is only helpful for differentiating between multiple `FirebaseAuth` instances.

I think the extra writes to the db shouldn't pose a big problem, and an added bonus is that this helps us with the "backfill" process.

File: app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt (hunk 1 above)
Comment (cyamonide) [reply app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt]: I could add `suspend`, but when we `upsert` on login I think we want the operation to be blocking.

Else we could run into the race condition where a user logs in to the app, and their user doesn't yet exist in the `users` collection, which could break the profile page or some other components.

I can make this async if there might be other use cases (other than login) for calling upsert.

Comment (aaronchlam) [reply app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt]: It's not actually blocking though. We're simply setting a callback function and then immediately returning.

File: app/src/main/java/com/uwcs446/socialsports/services/user/current/FirebaseCurrentUserRepository.kt (hunk 2 above)
Comment (aaronchlam) [reply app/src/main/java/com/uwcs446/socialsports/services/user/current/FirebaseCurrentUserRepository.kt:38]: Ya I think this is fine.

**Review by aaronchlam — DISMISSED** [review]

File: app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt (hunk 1 above)
Comment (cyamonide) [reply app/src/main/java/com/uwcs446/socialsports/services/user/FirebaseUserRepository.kt]: Pushed a fix in https://github.com/SPahooja/SocialSports/pull/53/commits/cafccb154c9cf412543186a1af04671da17cb96a - this uses `await()` and blocks at the main view level scope.

**Review by aaronchlam — DISMISSED** [review]

**Review by aaronchlam — APPROVED** [review]

**Pull request merged**